training.py

```diff
@@ -0,0 +1,226 @@
+        predictions = model.predictAll(valid_data).map(lambda r: ((r[0], r[1]), r[2]))
+        # get the rating result
+        ratesAndPreds = validation_data.map(lambda r: ((r[0], r[1]), r[2])).join(predictions)
+        # get the RMSE
+        MSE = ratesAndPreds.map(lambda r: (r[1][0] - r[1][1])**2).mean()
+        error = math.sqrt(MSE)
+        # add to errors
+        errors.append(error)
+
+    # plot
+    plt.figure(figsize=(12, 6))
+    plt.plot(arr_iters, errors)
+    plt.xlabel('number of iterations')
+    plt.ylabel('RMSE')
+    plt.title('ALS Learning Curve')
+    plt.grid(True)
+    plt.show()
+
+def get_movieId(df_movies, fav_movie_list,movies):
+    """
+    return all movieId(s) of user's favorite movies
+    
+    Parameters
+    ----------
+    df_movies: spark Dataframe, movies data
+    
+    fav_movie_list: list, user's list of favorite movies
+    
+    Return
+    ------
+    movieId_list: list of movieId(s)
+    """
+    movieId_list = []
+    for movie in fav_movie_list:
+        movieIds = df_movies \
+            .filter(movies.title.like('%{}%'.format(movie))) \
+            .select('movieId') \
+            .rdd \
+            .map(lambda r: r[0]) \
+            .collect()
+        movieId_list.extend(movieIds)
+    result = list(set(movieId_list))
+    #print(result)
+    return result
+
+def get_inference_data(train_data, df_movies, movieId_list):
+    """
+    return a rdd with the userid and all movies (except ones in movieId_list)
+
+    Parameters
+    ----------
+    train_data: spark RDD, ratings data
+
+    df_movies: spark Dataframe, movies data
+    
+    movieId_list: list, list of movieId(s)
+
+    Return
+    ------
+    inference data: Spark RDD
+    """
+    # get new user id
+    new_id = train_data.map(lambda r: r[0]).max() + 1
+    # return inference rdd
+    return df_movies.rdd \
+        .map(lambda r: r[0]) \
+        .distinct() \
+        .filter(lambda x: x not in movieId_list) \
+        .map(lambda x: (new_id, x))
+        
+def add_new_user_to_data(train_data, movieId_list, spark_context):
+    """
+    add new rows with new user, user's movie and ratings to
+    existing train data
+
+    Parameters
+    ----------
+    train_data: spark RDD, ratings data
+    
+    movieId_list: list, list of movieId(s)
+
+    spark_context: Spark Context object
+    
+    Return
+    ------
+    new train data with the new user's rows
+    """
+    # get new user id
+    new_id = train_data.map(lambda r: r[0]).max() + 1
+    # get max rating
+    max_rating = train_data.map(lambda r: r[2]).max()
+    # create new user rdd
+    user_rows = [(new_id, movieId, max_rating) for movieId in movieId_list]
+    new_rdd = spark_context.parallelize(user_rows)
+    
+    # return new train data
+    return train_data.union(new_rdd)
+
+
+def make_recommendation(movies,best_model_params, ratings_data, df_movies,fav_movie_list, n_recommendations, spark_context):
```

Comment:
What my team at HBO does is have this function(s) in a scoring.py file -> Essentially we split up the modeling layer into two components: 1) training and 2) scoring